### PR TITLE
[server][dvc] Lag arithmetic fix

### PR DESCRIPTION
**Pull request by FelixGV**
There is an edge case where an empty topic with an end offset of zero gets mistaken for an error code, and results in returning Long.MAX_VALUE as the lag instead.

## How was this PR tested?
It was not... tests should be added.

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 1]
```diff
@@ -2416,11 +2416,18 @@ protected long measureLagWithCallToPubSub(
       return Long.MAX_VALUE;
     }
     TopicManager tm = getTopicManager(pubSubServerName);
-    long endOffset = tm.getLatestOffsetCached(topic, partition) - 1;
+    long endOffset = tm.getLatestOffsetCached(topic, partition);
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2419]: what if we returned lag `0` if endOffset is `0`?

Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2419]: So... I guess the rationale is that if the end offset is zero, then it's an empty partition, and therefore there must be zero lag. We can do that if you think it makes the code clearer. Although the rest of the function would still be the same right? (i.e. we still the -1 later on, etc.)

Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2429]: I adopted your suggestion in the latest commit, and also added a unit test.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 2]
```diff
@@ -2412,15 +2412,36 @@ protected long measureLagWithCallToPubSub(
...
+      // -1 is a valid offset, which means that nothing was consumed yet, but anything below that is invalid.
       return Long.MAX_VALUE;
     }
-    TopicManager tm = getTopicManager(pubSubServerName);
-    long endOffset = tm.getLatestOffsetCached(topic, partition) - 1;
+    TopicManager tm = topicManagerProvider.apply(pubSubServerName);
+    long endOffset = tm.getLatestOffsetCached(topic, partition);
     if (endOffset < 0) {
+      // A negative value means there was a problem in measuring the end offset, and therefore we return "infinite lag"
       return Long.MAX_VALUE;
```
Comment (gaojieliu) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2432]: Shall we add a warn log here as it is unexpected?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 3]
```diff
@@ -2412,15 +2412,36 @@ protected long measureLagWithCallToPubSub(
...
+    } else if (endOffset == 0) {
```
Comment (sixpluszero) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2433]: Why do we have this branch? I think it can be merged into below branch. EndOffset = 0 then currentOffset must be -1? then below formula also applies

Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2433]: That is true. It was a suggestion from @sushantmane. It is not strictly necessary, but might be easier to read for some (maybe not all...?) people.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 4]
```diff
@@ -2412,15 +2412,36 @@ protected long measureLagWithCallToPubSub(
...
+      // The topic is empty and therefore, by definition, there cannot be any lag
+      return 0;
     }
-    return endOffset - currentOffset;
+
+    /**
+     * An empty topic should have an end offset of zero. A topic with a single message in it will have an end offset of
```
[the remark this answers is not in the document]
Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2439]: So I guess it would be more accurate to say that a topic with end offset of zero has always been empty. Whereas a topic which had messages in it before may now be empty (due either to time retention or to log compaction where all keys have gotten a null value [which we do not do, but anyway...]). In such case where a topic has had messages before but is empty now, the end offset would be > 0.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 2 above)
Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2432]: There is already an `ERROR` log in `TopicMetadataFetcher::getLatestOffsetCached` when returning the negative error code, so I would rather not add another one.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 3 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2433]: After thinking about this more, I think maybe we should remove this branch.  Sorry for the trouble

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 2 above)
Comment (gaojieliu) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2432]: SG

**Review by sushantmane — APPROVED** [review]
LGTM. Thanks, @FelixGV!

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 3 above)
Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:2433]: IMHO, I think it's fine to keep it, as it helps structure the comments to explain what each offset state means. Though the code could be one instruction shorter and achieve the same outcome, I don't think it matters in this case from a performance standpoint.

**Review by sixpluszero — APPROVED** [review]
